### ISSUE TITLE
Question about KVCache prefetch logic in LlamaFlashAttention2_AttnPred

### ISSUE THREAD

**Issue by OasisJiang** [MIRALab-USTC/LLM-AttentionPredictor#2]
Hello, I have been studying the code of this project and have some questions about the implementation logic in the `forward` function of `LlamaFlashAttention2_AttnPred`.

In the paper, you mention:

> “During the decoding phase, AttentionPredictor forecasts the critical token indices **S** for the next step. The framework then **prefetches the sparse KV cache with S** for the next step from the full cache to the GPU.”

However, I did not find any code in the project that actually performs **prefetching of the sparse KV cache**. Furthermore, the `time_series_predict` output — which should represent the important KV positions to prefetch — seems to be used only for generating `tsp_mask` and adding it to the **current** attention weights (`attn_weights`) that have already been computed.

This raises two questions:

1. **Why is the predicted mask for token *i+1* applied to the attention weights of the current token *i*?**
2. **Where is the logic that prefetches or extracts the sparse KVCache based on the predicted important indices?**

I hope I have not misunderstood the implementation, and I would greatly appreciate your clarification on this part of the code. Thank you!

**Comment by SimonSun0810** [comment]
I have the same questions. Could the author explain these?

**Comment by yezhaowei666** [comment]
me too

**Comment by lywang0** [comment]
I totally agree with @OasisJiang . It appears this codebase is merely a simulation to validate accuracy, but it is incapable of reproducing the latency/throughput metrics reported in the paper and it's not reasonable to use the code provided in the codebase to compare against other baselines.

**Comment by yangqy1** [comment]
@OasisJiang @SimonSun0810 @yezhaowei666 @lywang0 

Thank you for the detailed questions, and sincere apologies for the very late reply.

(1) In our implementation, **the prediction only uses information available up to step t−1 at the beginning of step t**, and the resulting mask is then used in step t. This is the same causal setup commonly used in prior work.

Concretely in `attnpred_llama_attention.py`, 
```
165            tsp_attn = self.time_series_predict(self.attn_history, token_id=key_states.shape[-2])
166            tsp_mask = self.create_tsp_mask(attn_weights_full, tsp_attn, topk=self.topk)
167            # self.recovery_rate = torch.sum(attn_weights_full[torch.where(tsp_mask==0)])/torch.sum(attn_weights_full)
168            attn_weights = attn_weights + tsp_mask
169            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
170           
171            if self.calibration_step is not None and self.decode_length % self.calibration_step == 0:
172                self.attn_history = self.update_attn_history(self.attn_history, attn_weights_full)
173            else:
174                self.attn_history = self.update_attn_history(self.attn_history, attn_weights)
```
we first run the predictor to obtain the mask for the current decoding step in line 165. **The prediction happens before the `attn_history` is updated with current attention**. Then, after the attention for the current step is computed, we update `attn_history` with the newly computed attention to serve as input for the next step’s prediction (line 172 and 174). 


(2) This repository currently contains only a simulation of prefetching for accuracy/behavior validation. The full prefetch implementation will be released as soon as possible (a few weeks).

Again, apologies for the delayed clarification, and thank you for taking the time to read the code so carefully.

**Comment by shuhao02** [comment]
Hi, I have tried to reproduce the latency/throughput in the paper, but the compute of the CNN is large and hard to eliminate (and grows with the context length). 

Do we have any way to reproduce the latency/throughput metrics reported in the paper?  I would greatly appreciate more clarification. Thank you!
